mysql-test/suite/innodb/t/instant_alter_charset.test

```diff
@@ -0,0 +1,527 @@
+
+
+create table latin1_swedish_special_case (
+  copy1 varchar(150) charset ascii collate ascii_general_ci,
+  copy2 char(150) charset ascii collate ascii_general_ci,
+  instant1 varchar(150) charset ascii collate ascii_general_ci,
+  instant2 char(150) charset ascii collate ascii_general_ci
+) engine=innodb;
+
+select c.name, c.prtype, c.mtype, c.len from information_schema.innodb_sys_columns as c inner join information_schema.innodb_sys_tables t on c.table_id = t.table_id
+  where t.name = 'test/latin1_swedish_special_case';
+alter table latin1_swedish_special_case
+  modify copy1 varchar(150) charset latin1 collate latin1_swedish_ci,
+  modify copy2 char(150) charset latin1 collate latin1_swedish_ci,
+  algorithm=copy;
+alter table latin1_swedish_special_case
+  modify instant1 varchar(150) charset latin1 collate latin1_swedish_ci,
+  modify instant2 char(150) charset latin1 collate latin1_swedish_ci,
+  algorithm=instant;
+select c.name, c.prtype, c.mtype, c.len from information_schema.innodb_sys_columns as c inner join information_schema.innodb_sys_tables t on c.table_id = t.table_id
+  where t.name = 'test/latin1_swedish_special_case';
+alter table latin1_swedish_special_case
+  modify copy1 varchar(150) charset latin1 collate latin1_general_ci,
+  modify copy2 char(150) charset latin1 collate latin1_general_ci,
+  algorithm=copy;
+alter table latin1_swedish_special_case
+  modify instant1 varchar(150) charset latin1 collate latin1_general_ci,
+  modify instant2 char(150) charset latin1 collate latin1_general_ci,
+  algorithm=instant;
+select c.name, c.prtype, c.mtype, c.len from information_schema.innodb_sys_columns as c inner join information_schema.innodb_sys_tables t on c.table_id = t.table_id
+  where t.name = 'test/latin1_swedish_special_case';
+
+drop table latin1_swedish_special_case;
+
+create table regression (a varchar(100) charset utf8mb3 primary key, b int) engine=innodb;
+alter table regression convert to character set utf8mb4;
+drop table regression;
+
+# ROW_FORMAT=DYNAMIC limitation:
+# size in bytes cannot be increased from less of equal that 255 to more than 255
+create table boundary_255 (
+  a varchar(50) charset ascii,
+  b varchar(200) charset ascii,
+  c varchar(300) charset ascii
+) engine=innodb;
+
+alter table boundary_255
+  modify a varchar(50) charset utf8mb3,
+  algorithm=instant;
+
+--error ER_ALTER_OPERATION_NOT_SUPPORTED_REASON
+alter table boundary_255
+  modify b varchar(200) charset utf8mb3,
+  algorithm=instant;
+
+alter table boundary_255
+  modify c varchar(300) charset utf8mb3,
+  algorithm=instant;
+
+drop table boundary_255;
+
+create table fully_compatible (
+  id int auto_increment unique key,
+  from_charset char(255),
+  from_collate char(255),
+  to_charset char(255),
+  to_collate char(255)
+);
+
+insert into fully_compatible (from_charset, from_collate, to_charset, to_collate) values
+  ('utf8mb3', 'utf8mb3_general_ci',           'utf8mb4', 'utf8mb4_general_ci'),
```

Comment:
Could we declare `from_charset` and `to_charset` virtual columns, so that the `INSERT` statement can only populate the collations?
Can we use `CREATE TEMPORARY TABLE` for this metadata?